### ISSUE TITLE
F# compiler is broken with latest CoreCLR runtime

### ISSUE THREAD

**Issue by eerhardt** [dotnet/coreclr#13919]
Trying to take the latest 2.1 runtime into dotnet/cli is currently blocked because the F# compiler is running into what appears to be a JIT issue.

See https://github.com/dotnet/cli/pull/7606#issuecomment-328694723 for my analysis of debugging the issue.

To repro:

1. Unzip [JITBug.zip](https://github.com/dotnet/coreclr/files/1296754/JITBug.zip) to a folder.
1. Set your CWD into the `JITBug` folder you extracted
1. Using the [latest shared runtime](https://dotnetcli.blob.core.windows.net/dotnet/Runtime/master/dotnet-runtime-latest-win-x64.zip) execute:
`dotnet.exe fsc.exe -o:obj\Debug\netcoreapp2.0\FSharpTest.dll -g --debug:portable --noframework --optimize- -r:refs\FSharp.Core.dll -r:refs\System.Runtime.dll --target:exe --warn:3 --fullpaths --flaterrors --highentropyva- --targetprofile:netcore --simpleresolution --nocopyfsharpcore .NETCoreApp,Version=v2.0.AssemblyAttributes.fs Program.fs`

You will get an error like:

```
FailFast: Unknown exception seen in compiler: System.InvalidCastException: Unable to cast object of type 'Ident' to type 'LongIdentWithDots'.
   at Microsoft.FSharp.Core.LanguagePrimitives.IntrinsicFunctions.UnboxGeneric[T](Object source)
   at Microsoft.FSharp.Compiler.Parser._fsyacc_reductions@14808-1158.Invoke(IParseState parseState)
   at Internal.Utilities.Text.Parsing.Implementation.interpret[tok](Tables`1 tables, FSharpFunc`2 lexer, LexBuffer`1 lexbuf, Int32 initialState)
   at Microsoft.FSharp.Compiler.CompileOps.ParseInput(FSharpFunc`2 lexer, ErrorLogger errorLogger, LexBuffer`1 lexbuf, FSharpOption`1 defaultNamespace, String filename, Tuple`2 isLastCompiland)
   at Microsoft.FSharp.Compiler.CompileOps.ParseOneInputLexbuf@3575.Invoke(LexBuffer`1 lexbuf)
   at Microsoft.FSharp.Compiler.Lexhelp.reusingLexbufForParsing[a](LexBuffer`1 lexbuf, FSharpFunc`2 f)
```

/cc @jkotas @briansull

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
Seems likely it is one of my changes, so I'll take a look if @briansull hasn't already started in on this.

**Comment by briansull** [comment]
@AndyAyersMS I haven't yet started debug this

**Comment by jkotas** [comment]
It seems to be caused by https://github.com/dotnet/coreclr/pull/13314

**Comment by jkotas** [comment]
@JosephTremoulet Could you please take a look?

**Comment by JosephTremoulet** [comment]
Sure.

**Comment by JosephTremoulet** [comment]
I have now put up a PR for a fix (#13957).
